### PR TITLE
[Raiden Weight Sync 4/6] Manage cache lifecycle and prefix caching in RaidenWeightSyncDelegate

### PR DESCRIPTION
**Pull request by YixuanWang-99**
## Overview
Part of the stacked Raiden weight-sync enablement PR chain replacing #2083.

**Stack:**
- [T1] https://github.com/google/tunix/pull/2146: Multi-host discovery & multi-axis sharding in RaidenHandler
- [T2a] https://github.com/google/tunix/pull/2147: Transport selection, FFI preflight mismatch check, and host-stage normalization
- [T3] https://github.com/google/tunix/pull/2148: Extract RaidenDestinationWeightSyncMixin
- **[T4] google/tunix (this PR)**: Cache lifecycle and prefix caching
- [T5] google/tunix: MoE 128-lane interleaving for TPU GMM layout (pairs with MaxText M2)
- [T6] google/tunix: MaxText trainer configuration plumbing

## Details
- Resets prefix cache and clears stale KV entries upon weight transfer.
- Adds `--enable_prefix_caching` flag (default `false`) to avoid recurrent state desync in hybrid models (such as Qwen3.5 GDN layers).

### CODE REVIEW

File: tunix/experimental/weight_sync/raiden_weight_sync_delegate.py [hunk 1]
```diff
@@ -36,14 +40,33 @@ class RaidenWeightSyncDelegate:
   abort after a partial weight_sync cannot restore the previous weights.
   """
 
-  def __init__(self, *args, worker_index: int = 0, **kwargs):
-    super().__init__(*args, **kwargs)
+  def __init__(
+      self,
+      *args,
+      job_name: Optional[str] = None,
```
Comment (tianshub) [inline tunix/experimental/weight_sync/raiden_weight_sync_delegate.py]: you don't need both job_name, server_id and worker_index as they serve duplicated purpose. Just rely on worker_index and compose the job name same as other PR

File: tunix/experimental/weight_sync/raiden_weight_sync_delegate.py [hunk 2]
```diff
@@ -36,14 +40,33 @@ class RaidenWeightSyncDelegate:
...
+      **kwargs,
+  ):
+    del args, kwargs
+    # Raiden partitions the weights across every unit sharing a job_name, so
+    # replicas that all call themselves "rollout" get a slice each instead of a
+    # copy each. server_id is already unique per replica and shared across the
+    # hosts within one, which is exactly the grouping job_name needs.
+    self.server_id = server_id
+    self.job_name = job_name or server_id or "rollout"
+    self._sampler = sampler
```
Comment (tianshub) [inline tunix/experimental/weight_sync/raiden_weight_sync_delegate.py]: I don't see how this sampler is used?

File: tunix/experimental/weight_sync/raiden_weight_sync_delegate.py [hunk 3]
```diff
@@ -36,14 +40,33 @@ class RaidenWeightSyncDelegate:
...
     self._synchronizers: List[Any] = [
         raiden_synchronizer.RaidenSynchronizer(
-            "rollout", worker_index=worker_index, auto_h2d=True
+            self.job_name,
+            worker_index=worker_index,
+            auto_h2d=True,
         )
     ]
     self._version = 0
+    self._sync_lock = asyncio.Lock()
```
Comment (tianshub) [inline tunix/experimental/weight_sync/raiden_weight_sync_delegate.py]: do you expect multiple samplers in one worker? if not I don't think we need this lock

File: tunix/experimental/weight_sync/raiden_weight_sync_delegate.py (hunk 2 above)
Comment (wang2yn84) [reply tunix/experimental/weight_sync/raiden_weight_sync_delegate.py]: It's used to invoke reset_prefix_cache and delete kv cache.

File: tunix/experimental/weight_sync/raiden_weight_sync_delegate.py [hunk 4]
```diff
@@ -36,14 +40,33 @@ class RaidenWeightSyncDelegate:
...
+      sampler: Optional[Any] = None,
+      **kwargs,
+  ):
+    del args, kwargs
+    # Raiden partitions the weights across every unit sharing a job_name, so
+    # replicas that all call themselves "rollout" get a slice each instead of a
+    # copy each. server_id is already unique per replica and shared across the
+    # hosts within one, which is exactly the grouping job_name needs.
+    self.server_id = server_id
+    self.job_name = job_name or server_id or "rollout"
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/raiden_weight_sync_delegate.py]: Server id is much worse than "rollout". How can user infer if certain server id is rollout or trainer?

File: tunix/experimental/weight_sync/raiden_weight_sync_delegate.py (hunk 1 above)
Comment (wang2yn84) [reply tunix/experimental/weight_sync/raiden_weight_sync_delegate.py]: The only reason why we need worker_index here is because there will be multiple rollout servers. That's the same reason why we don't pass job_name and instead using hardcoded "rollout" for simplicity. Please revert it back if it's not necessary.

File: tunix/experimental/weight_sync/raiden_weight_sync_delegate.py [hunk 5]
```diff
@@ -69,30 +94,101 @@ async def get_weight_sync_metadata(self, **kwargs) -> Any:
...
+      if sampler is not None:
+        model_runner = getattr(sampler, "_model_runner", None)
+        if model_runner is not None and hasattr(model_runner, "state_leaves"):
+          state = getattr(model_runner, "state", None)
+          if state is not None:
+            model_runner.state_leaves = tuple(
+                jax.tree_util.tree_leaves(state)
+            )
+
+        llm = getattr(sampler, "llm", None)
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/raiden_weight_sync_delegate.py]: These are the impl details of sampler. Consider move these out of update_param() to a new API delete_cache() in sampler so that you only need to call self.sampler.delete_cache()?

File: tunix/experimental/weight_sync/raiden_weight_sync_delegate.py [hunk 6]
```diff
@@ -69,30 +94,101 @@ async def get_weight_sync_metadata(self, **kwargs) -> Any:
...
-    del sync_request, kwargs
-    return True
+    del kwargs
+    async with self._sync_lock:
+      if self._has_round(sync_request):
+        if not self._tracker.admit(sync_request, "prepared"):
+          return True
+        self._tracker.complete(sync_request, "prepared")
+
+      sampler = getattr(self, "_sampler", None)
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/raiden_weight_sync_delegate.py]: Instead of defensive coding like this, you probably should check the existence of sampler and error out if sampler is not available. Then you can just call self.sampler.xxx() everywhere.

File: tunix/experimental/weight_sync/raiden_weight_sync_delegate.py (hunk 1 above)
Comment (A9isha) [reply tunix/experimental/weight_sync/raiden_weight_sync_delegate.py]: Done. I've removed the job_name and server_id arguments from __init__ and reverted to relying purely on worker_index with the hardcoded "rollout" job name.

File: tunix/experimental/weight_sync/raiden_weight_sync_delegate.py (hunk 2 above)
Comment (A9isha) [reply tunix/experimental/weight_sync/raiden_weight_sync_delegate.py]: yes it's used to trigger the KV cache clearing and reinitialization during the weight sync lifecycle. Following up on the other review comments, I've also refactored how it's used so we just call a clean `self._sampler.delete_cache()` API rather than leaking the vLLM implementation details here.

File: tunix/experimental/weight_sync/raiden_weight_sync_delegate.py (hunk 3 above)
Comment (A9isha) [reply tunix/experimental/weight_sync/raiden_weight_sync_delegate.py]: SG - for now, we only expect one sampler per worker right now, so I've removed _sync_lock (and the asyncio import) entirely. I also left a TODO(tunix-dev) to reintroduce a lock if we ever support multiple samplers per worker in the future.

File: tunix/experimental/weight_sync/raiden_weight_sync_delegate.py (hunk 5 above)
Comment (A9isha) [reply tunix/experimental/weight_sync/raiden_weight_sync_delegate.py]: Done! I've extracted the cache and state_leaves logic into new delete_cache() and reinitialize_cache() methods directly on VllmSampler (with no-op defaults in BaseSampler). The RaidenWeightSyncDelegate now just calls these unified APIs cleanly, and update_params() reuses them as well.

File: tunix/experimental/weight_sync/raiden_weight_sync_delegate.py (hunk 6 above)
Comment (A9isha) [reply tunix/experimental/weight_sync/raiden_weight_sync_delegate.py]: Done. I've replaced the defensive getattr checks with an explicit if self._sampler is None: raise RuntimeError(...).

File: tunix/experimental/weight_sync/raiden_weight_sync_delegate.py (hunk 4 above)
Comment (A9isha) [reply tunix/experimental/weight_sync/raiden_weight_sync_delegate.py]: yes hardcoded this to "rollout"

File: tunix/experimental/weight_sync/raiden_weight_sync_delegate.py [hunk 7]
```diff
@@ -36,14 +39,25 @@ class RaidenWeightSyncDelegate:
...
-  def __init__(self, *args, worker_index: int = 0, **kwargs):
-    super().__init__(*args, **kwargs)
+  def __init__(
+      self,
+      *args,
+      worker_index: int = 0,
+      sampler: Optional[Any] = None,
+      **kwargs,
+  ):
+    del args, kwargs
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/raiden_weight_sync_delegate.py:47]: Just fyi, I'm having a small cl/979409558 to fix the broken Tunix head. Let's try to land this PR first if possible so that you don't have to rebase again. But if this PR still needs iterations, I might land that cl first.

File: tunix/experimental/weight_sync/raiden_weight_sync_delegate.py [hunk 8]
```diff
@@ -52,10 +66,12 @@ def is_bounded(
     return all(s.bound for s in self._synchronizers)
 
   async def bind_weight_sync(
-      self, sync_request: Any = None, state: Any = None, **kwargs
+      self, sync_request: Any = None, state: Any = None, sampler: Any = None, **kwargs
   ) -> Any:
     """Binds destination-side transport resources for weight sync."""
     del sync_request, kwargs
+    if sampler is not None:
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/raiden_weight_sync_delegate.py:71]: Why do we pass sampler in again when sampler is passed in init?

Comment (A9isha) [reply tunix/experimental/weight_sync/raiden_weight_sync_delegate.py:71]: Great point! In the actual rollout adapters, the delegate is instantiated before the sampler is fully ready, so we were only passing it to `__init__` in the tests anyway. I've removed sampler from `__init__` entirely to avoid confusion and updated the tests to pass it via `bind_weight_sync()` exactly like the real code does.

Comment (wang2yn84) [reply tunix/experimental/weight_sync/raiden_weight_sync_delegate.py:71]: In the future, we might want to have set_samper() for clearer API.
